py-gevent: @:23.9.0 conflicts with py-cython@3.0.10 (#45257)

jaguillette opened the pull request:
Adds a conflict statement for py-cython@3.0.10, which fails to build py-gevent.
@spackbot fix style
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

jaguillette committed:
Add conflict for py-cython@3.0.10 (#45257)

Adds a conflict statement for py-cython@3.0.10, which fails to build py-gevent.

wdconinc commented on var/spack/repos/builtin/packages/py-gevent/package.py:
```diff
@@ -41,6 +41,8 @@ class PyGevent(PythonPackage):
     # https://github.com/gevent/gevent/issues/1599
     conflicts("^py-cython@3:", when="@:20.5.0")
+    # https://github.com/gevent/gevent/issues/2031
+    conflicts("^py-cython@3.0.10", msg="py-gevent fails to build when using cython@3.0.10")
```
This seems too broad. Should this apply to all `py-gevent` versions?

jaguillette replied:
The issue in the gevent repo (https://github.com/gevent/gevent/issues/2031) reports having tested with gevent versions 23.9.1 and 24.2.1. It's not clear to me whether an older version of gevent would work with the newer cython, or when a patch would be expected upstream in gevent.

wdconinc replied:
One reason to say it is too broad is that I can't reproduce this on 24.2.1:
```
==> Installing py-cython-3.0.10-okmooj2yjdesogia5riggblwbsufng3w [40/44]
==> No binary for py-cython-3.0.10-okmooj2yjdesogia5riggblwbsufng3w found: installing from source
==> Using cached archive: /opt/spack/cache/_source-cache/archive/dc/dcc96739331fb854dcf503f94607576cfe8488066c61ca50dfd55836f132de99.tar.gz
==> No patches needed for py-cython
==> py-cython: Executing phase: 'install'
==> py-cython: Successfully installed py-cython-3.0.10-okmooj2yjdesogia5riggblwbsufng3w
  Stage: 0.11s.  Install: 6m 10.59s.  Post-install: 0.66s.  Total: 6m 11.60s
[+] /opt/software/linux-ubuntu24.04-skylake/gcc-13.2.0/py-cython-3.0.10-okmooj2yjdesogia5riggblwbsufng3w
(snip)
==> Installing py-gevent-24.2.1-6dt6avm6anucbombpy6bdecmokl6ic3e [44/44]
==> No binary for py-gevent-24.2.1-6dt6avm6anucbombpy6bdecmokl6ic3e found: installing from source
==> Using cached archive: /opt/spack/cache/_source-cache/archive/43/432fc76f680acf7cf188c2ee0f5d3ab73b63c1f03114c7cd8a34cebbe5aa2056.tar.gz
==> No patches needed for py-gevent
==> py-gevent: Executing phase: 'install'
==> py-gevent: Successfully installed py-gevent-24.2.1-6dt6avm6anucbombpy6bdecmokl6ic3e
  Stage: 0.15s.  Install: 3m 19.73s.  Post-install: 0.78s.  Total: 3m 20.85s
[+] /opt/software/linux-ubuntu24.04-skylake/gcc-13.2.0/py-gevent-24.2.1-6dt6avm6anucbombpy6bdecmokl6ic3e
```

jaguillette replied:
Hm, interesting. Does 23.9.1 error out using `py-cython@3.0.10` for you? If it does, then the conflict could be set for all versions before 24.2.1, since that's the release after 23.9.1 [on pypi](https://pypi.org/project/gevent/#history). I'll try 24.2.1 on our system once I get the chance. Thanks!

jaguillette replied:
I was able to do a `py-gevent@23.9.1` build with `py-cython@3.0.10`, so that seems like an upper bound for the conflict. I've added that constraint to the PR. 

I did add the checksums for newer versions manually, and I'm not sure what the normal process is for adding versions of packages. I haven't done testing beyond "does it install", and this issue came out of a package that has `py-gevent` as a dependency, so I'm not familiar with `py-gevent` itself either.

wdconinc replied:
> I'm not sure what the normal process is for adding versions of packages

PRs are indeed the way to go :-) For python packages, look additionally for changes to dependencies in pyproject.toml (or setup.cfg|py). Reviewers will check hashes and do a double check of dependency changes. CI will do style and consistency checks.

jaguillette replied:
I've updated the dependency spec (thanks for the fix on that!), and if we end up wanting to add new versions I'll open a separate PR. Sound good?

bernhardkaindl replied:
As far as I can see, the `when=` condition is added properly and this conversation is recolved.

jaguillette committed:
Add version constraint to conflict

I tried an install with v23.9.1 of `py-gevent` and it installed successfully with `cython@3.0.10`, so that seems to be the upper bound where such a patch would be needed.

wdconinc commented on var/spack/repos/builtin/packages/py-gevent/package.py:
```diff
@@ -41,6 +41,8 @@ class PyGevent(PythonPackage):
     # https://github.com/gevent/gevent/issues/1599
     conflicts("^py-cython@3:", when="@:20.5.0")
+    # https://github.com/gevent/gevent/issues/2031
+    conflicts("^py-cython@3.0.10", when="@:23.9.1", msg="py-gevent fails to build when using cython@3.0.10")
```
Upper limits on version ranges are inclusive in spack, so this would need something like
```suggestion
    conflicts("^py-cython@3.0.10", when="@:23.9.0", msg="py-gevent fails to build when using cython@3.0.10")
```
(That is if it fails to build with 23.9.0 but succeeds with 23.9.1.)

jaguillette committed:
fix version constraint

would have included a functioning version in commit range

Co-authored-by: Wouter Deconinck <wdconinc@gmail.com>

jaguillette committed:
Fix style issue

jaguillette committed:
style fix

bernhardkaindl reviewed (approved):
The `when=` condition is added properly, both conversations resolved. Built on Ubuntu 22.04
